Adding workspace types endpoints

ulferts opened the pull request:
# Ticket

https://community.openproject.org/wp/67278

# What are you trying to accomplish?

Add `Portfolio` and `Program` as resources of their own next to `Project`. There is also then the abstract `Workspace` that is introduced as a concept in the api.

* [x] Add index endpoints with full capability of the typical OP index endpoint(i.e. filtering, ordering, signaling, pagintion, ...)
  * `GET api/v3/workspaces` - returns instances of the three other types
  * `GET api/v3/portfolios`
  * `GET api/v3/programs`
* [x] Add show endpoints for portfolios and programs
  * `GET api/v3/portfolios/:id`
  * `GET api/v3/programs/:id`
* [x] The two added endpoints for individual resources, same as `api/v3/projects/:id` have additional endpoints for
  * `DELETE api/v3/portfolios/:id` and `DELETE api/v3/programs/:id` 
  * `PATCH api/v3/portfolios/:id` and `PATCH api/v3/programs/:id` 
  * `POST api/v3/portfolios/:id/form` and `PATCH api/v3/programs/:id/form` 
* [x] Adds an endpoint for the `Workspace` schema
  * `GET api/v3/workspaces/schema`  
  * `GET api/v3/projects/schema` is kept for backwards compatibility but references will point to the workspace schema
* [x] Document added routes
* [x] Move `projects/schema` to `workspaces/schema`
  * [x] Change type in "parent" property
  * [x] Change `self` link
* [x] Drive by fix that became entangled with the work:
  * align favorit/favorited/disfavor with how it is done for staring of queries. It becomes a property and action links on the workspace/project representer. That approach is then also documented in the api docs.

# Out of scope

* Creating portfolios and programs via the API
* Changes to the representers (e.g. versions and memberships) that have project links.
* Move auxiliary endpoints scoped before under `api/v3/projects/:id` to be under `api/v3/workspaces/:id`. E.g. `api/v3/workspaces/:id/available_assignees`. Done in #21148

ulferts committed:
adding workspace types endpoints

EinLama commented on docs/api/apiv3/paths/portfolio.yml:
```diff
@@ -0,0 +1,221 @@
+  description: |-
+    Deletes the portfolio permanently. As this is a lengthy process, the actual deletion is carried out asynchronously.
+    So the project might exist well after the request has returned successfully. To prevent unwanted changes to
+    the project scheduled for deletion, it is archived at once.
```
```suggestion
    So the portfolio might exist well after the request has returned successfully. To prevent unwanted changes to
    the portfolio scheduled for deletion, it is archived at once.
```

EinLama commented on docs/api/apiv3/paths/portfolios.yml:
```diff
@@ -0,0 +1,95 @@
+# /api/v3/portfolios
+---
```
To access /portfolios in the web app, the feature flag must be enabled.

As for permissions, the global permission `add_portfolios` or any `view_project` permission in an active portfolio is required.

There is no mention of this here - or in the programs-sibling. Dropping this comment so that I don't forget to pick up on it later.

EinLama replied:
I could also not find where flag and permission are checked.

ulferts replied:
I applied the feature flag check to the portfolio(s)/program(s) endpoints. But I did not add the permission checks as I believe the use case to be different here. Checking those permission in the UI prevents the user from unnecessarily seeing the module without being able to do anything about it. That does not apply to the api. And in both cases, users will only see results in case they have any permission in the portfolio/programs so the check for `view_project` is ingrained into the queries already.

ulferts replied:
What needs to be added eventually to this API is a check for the enterprise edition.

EinLama replied:
True, good point!

EinLama commented on docs/api/apiv3/paths/workspaces.yml:
```diff
@@ -0,0 +1,95 @@
+    provided filter, the result set is always limited to only contain workspace the
+    client is allowed to see.
+
+    Since workflows are the generic term for a number of resources like projects and portfolios, the returned
```
These should be pluralized to stay consistent with the specific workspace type resources.

Further below, there is a word mismatch and workflow is used where workspace should have been:

```suggestion
  summary: List workspaces
  operationId: list_workspaces
  tags:
    - Workspaces
  description: Returns a collection of workspaces. The collection can be filtered via
    query parameters similar to how work packages are filtered. In addition to the
    provided filter, the result set is always limited to only contain workspaces the
    client is allowed to see.

    Since workspaces are the generic term for a number of resources like projects and portfolios, the returned
```

EinLama commented on docs/api/apiv3/paths/workspaces.yml:
```diff
@@ -0,0 +1,95 @@
+        + parent_id: filters workspace by their parent.
+        + principal: based on members of the portfolio.
+        + project_phase_any: based on the project phases active in a portfolio.
+        + project_status_code: based on status code of the portfolio
```
portfolio -> workspaces


```suggestion
        + active: based on the active property of the workspace
        + ancestor: filters workspaces by their ancestor. A workspace is not considered to be its own ancestor.
        + available_project_attributes: filters workspaces based on the activated project attributes.
        + created_at: based on the time the portfolio was created
        + favorited: based on the favorited property of the workspace
        + id: based on workspace id.
        + latest_activity_at: based on the time the last activity was registered on a workspace.
        + name_and_identifier: based on both the name and the identifier.
        + parent_id: filters workspace by their parent.
        + principal: based on members of the workspace.
        + project_phase_any: based on the project phases active in a workspace.
        + project_status_code: based on status code of the workspace
```

EinLama commented on docs/api/apiv3/paths/workspaces.yml:
```diff
@@ -0,0 +1,95 @@
+        + storage_id: filters workspace by linked storages
+        + storage_url: filters workspace by linked storages identified by the host url
+        + type_id: based on the types active in a portfolio.
+        + user_action: based on the actions the current user has in the portfolio.
```
```suggestion
        + type_id: based on the types active in a workspace.
        + user_action: based on the actions the current user has in the workspace.
```

EinLama commented on docs/api/apiv3/paths/workspaces.yml:
```diff
@@ -0,0 +1,95 @@
+        static but rather dynamically created on each OpenProject instance, a list cannot be provided. Those filters follow the schema:
+        + project_start_gate_[id]: a filter on a project phase's start gate active in a portfolio. The id is the id of the phase the gate belongs to.
+        + project_finish_gate_[id]: a filter on a project phase's finish gate active in a portfolio. The id is the id of the phase the gate belongs to.
+        + project_phase_[id]: a filter on a project phase active in a portfolio. The id is the id of the phase queried for.
```
```suggestion
        + project_start_gate_[id]: a filter on a project phase's start gate active in a workspace. The id is the id of the phase the gate belongs to.
        + project_finish_gate_[id]: a filter on a project phase's finish gate active in a workspace. The id is the id of the phase the gate belongs to.
        + project_phase_[id]: a filter on a project phase active in a workspace. The id is the id of the phase queried for.
```

EinLama commented on docs/api/apiv3/paths/workspaces_schema.yml:
```diff
@@ -0,0 +1,156 @@
+                _dependencies: []
+                _attributeGroups:
+                  - _type: ProjectFormCustomFieldSection
+                    name: Project Details
```
Should probably use workspace in the example names? Not sure about this one, as users will never create/edit actual, generic workspaces. So for this case, an example using any of these types might be better 🤔 But the other examples in this file use `workspace`, so to stay consistent:


```suggestion
                    name: Workspace Details
```

EinLama commented on docs/api/apiv3/tags/portfolios.yml:
```diff
@@ -0,0 +1,56 @@
+---
+description: |-
+  Portfolios are one of the types of [workspaces](https://www.openproject.org/docs/api/endpoints/workspaces) in OpenProject structuring the information (e.g. work packages, wikis) into smaller sets. They are typically used to group, structure and manage programs and projects that 
```
```suggestion
  Portfolios are one of the types of [workspaces](https://www.openproject.org/docs/api/endpoints/workspaces) in OpenProject structuring the information (e.g. work packages, wikis) into smaller sets. They are typically used to group, structure and manage programs, projects and other portfolios that 
```

EinLama replied:
should include the option of sub-portfolios

EinLama commented on docs/api/apiv3/tags/programs.yml:
```diff
@@ -0,0 +1,58 @@
+---
+description: |-
+  Programs are one of the types of [workspaces](https://www.openproject.org/docs/api/endpoints/workspaces) in OpenProject structuring the information (e.g. work packages, wikis) into smaller groups. They can be used in a classic program management approach but also when structuring work by departments.
```
duplicated paragraph

EinLama commented on docs/api/apiv3/components/examples/portfolio_collection.yml:
```diff
@@ -0,0 +1,40 @@
+  offset: 1
+  _embedded:
+    elements:
+      - _hint: Portfolio resource shortened for brevity
+        id: 1
+        identifier: initialportfolio
+        name: DeathStar construction
+        active: true
+        public: true
+      - _hint: Portfolio resource shortened for brevity
+        id: 2
+        identifier: mysecret
+        name: Palpatine's secret plan
+        active: true
+        public: false
```
This example data sounds like a better fit for a project than for a portfolio.

Maybe we could use our figma dummy text here instead:

> Digital Connect 2030
> This collection showcases a diverse selection of projects that highlight our commitment to excellence and our passion for design. From innovative web applications to stunning graphic designs, each piece reflects our unique approach and dedication to delivering impactful solutions. Discover the journey of our work and see how we bring ideas to life.

ulferts replied:
I didn't use the description here but applied it to the single portfolio example. Here, I only used different titles.

EinLama reviewed (changes requested):
Found some typos and slight wording inconsistencies. And there is a potential issue about checking portfolio permissions and the feature flag for portfolios and programs.

Thanks for adding the favorited routes and for splitting up the project factory into different files 👏

ulferts committed:
fix and adapt api docs

ulferts committed:
apply feature flag to portfolio and program api

ulferts commented:
@EinLama thanks for the review. 

I addressed your comments and changed the API documentation as wished. The feature flag check is added as well. I didn't add another permission check as I deem that unnecessary (please see below for the why). 

 I had to force push to rebase the branch. The original big commit is unchanged. The changes are solely in the two new commits so only those need to be reviewed.

EinLama reviewed (approved):
Nice, LGTM 👏
